iroha_p2p/src/network.rs

```diff
@@ -0,0 +1,154 @@
+                        yield NewPeerResult::Err { error };
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    async fn connect_peer(&mut self, id: PeerId, addr: Addr<Network<K, E>>) {
+        let peer = Peer::new(id.clone(), None, State::Connecting, addr)
+            .start()
+            .await;
+        drop(self.peers.insert(id, peer));
+    }
+}
+
+#[async_trait::async_trait]
+impl<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static> Actor for Network<K, E> {
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        // from peer
+        self.broker.subscribe::<ReceivedPeerMsg, _>(ctx);
+        // from other iroha subsystems
+        self.broker.subscribe::<SendPeerMsg, _>(ctx);
+        // from listener
+        ctx.notify_with_context(self.listener_stream());
+    }
+}
+
+#[async_trait::async_trait]
+impl<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static> ContextHandler<Connect>
+    for Network<K, E>
+{
+    type Result = ();
+
+    async fn handle(&mut self, ctx: &mut Context<Self>, msg: Connect) {
+        let addr = ctx.addr();
+        self.connect_peer(msg.id, addr).await;
+    }
+}
+
+#[async_trait::async_trait]
+impl<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static> Handler<SendPeerMsg>
+    for Network<K, E>
+{
+    type Result = ();
+
+    async fn handle(&mut self, msg: SendPeerMsg) {
+        let addr: &Addr<Peer<K, E>> = &self.peers[&msg.id];
+        addr.do_send(msg).await;
+    }
+}
+
+#[async_trait::async_trait]
+impl<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static> Handler<ReceivedPeerMsg>
+    for Network<K, E>
+{
+    type Result = ();
+
+    async fn handle(&mut self, _msg: ReceivedPeerMsg) {
+        // TODO send message to Torii
+        //self.handler.do_send(msg.message);
+    }
+}
+
+#[async_trait::async_trait]
+impl<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static>
+    ContextHandler<NewPeerResult> for Network<K, E>
+{
+    type Result = ();
+
+    async fn handle(&mut self, ctx: &mut Context<Self>, peer: NewPeerResult) {
+        let addr = ctx.addr();
+        match peer {
+            NewPeerResult::Ok { stream, id } => {
+                let addr = Peer::new(id.clone(), Some(stream), State::ConnectedFrom, addr)
+                    .start()
+                    .await;
+                drop(self.peers.insert(id, addr));
+            }
+            NewPeerResult::Err { error } => {
+                warn!("Error in listener! {}", error);
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, iroha_actor::Message)]
+pub struct Connect {
+    pub id: PeerId,
+}
+
+#[derive(Clone, Debug, iroha_actor::Message)]
+pub struct ReceivedPeerMsg {
+    pub data: Vec<u8>,
+    pub id: PeerId,
+}
+
+#[derive(Clone, Debug, iroha_actor::Message)]
+pub struct SendPeerMsg {
+    pub data: Vec<u8>,
+    pub id: PeerId,
+}
```

Comment:
Could you abstract this over Data?

```rust
pub struct SendPeer<T: Encode> {
    pub data: T,
    pub id: PeerId,
}

impl<T: Encode> Handler<SendPeer<T>> for Network {
...
}
```

Reply:
Same for `ReceivedPeer` but `T: Decode`

Reply:
No, the serialization will be out of the `Network`.

Reply:
Why?